### DISCUSSION TITLE
How to find the instance name of a performance counter when running under IIS

### DISCUSSION THREAD

**Discussion by tronda** [oracle/dotnet-db-samples#299] [Q&A]
I have a .NET Framework ASP.NET WebAPI running under IIS. For this application I would like it to report performance counter values as metrics to an OpenTelemetry Collector. The challenge is how I can construct the instance name at runtime when running under IIS. I see from the [simple sample](https://github.com/oracle/dotnet-db-samples/blob/f5bd7cf6d7d868a049bf5339aa1bd2f9f067cc32/samples/performance-counters/ODP-perf-counters.cs#L59) how this can be done, but I don't see how I can achieve the same from a running WebAPI application (one among several other WebAPI applications on the same machine). 

When I open PerfMon in Windows I see the instance name being similar to this:
`_lm_w3svc_1_root_aaaa_bbbb-w...[12100,2]`. I have not found a way to construct this instance name at runtime.

**Comment by alexkeh** [comment]
This question may be more applicable to an ASP.NET forum to ask in terms of how others have identified the specific web app instance from PerfMon.

I suppose you could retrieve the instance of every web app running, then check whether it has the ODP.NET counter being published. If so, you've found the instance you're looking for.

**Reply by tronda** [reply]
We have a .NET 4.8 application running under IIS which utilizes OpenTelemetry manual instrumentation. We would like to be able to push the `NumberOfActiveConnections` and `NumberOfFreeConnections` as metrics to the OpenTelemetry Collector according to the [database semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/database-metrics.md#connection-pools). Doing so has turned out to be a bit difficult. The ApplicationPool user that runs the application has to be part of the Performance Monitoring group in order to pull performance counters. Also I've been struggling with getting the instance name correct - mainly related to the process id which is part of the instance name. 
Because of these issues I've been thinking if I can solve this issue with a different approach. Oracle has recently published a new OpenTelemetry nuget package, but according to the documentation, it only focuses on traces. Will this package in the future also push metrics (particularly related to db-connections) to an OpenTelemetry Collector or are this not part of the scope? 
When it comes to .NET Core processes running under Linux - how can I get connection related information from the ODP.NET Core Managed driver? Any plans for supporting that? Have you considered exposing an API so one can get this information without going through performance counters?

**Comment by alexkeh** [comment]
The Oracle OpenTelemetry provider will first focus on building out trace functionality. 

For .NET counters, the ODP.NET team is working on supporting System.Diagnostics.Metrics in an upcoming release that will work for .NET Core. Once that's done, there may not be a need for a specific OpenTelemetry ODP.NET metrics implementation as .NET Metrics integrates with OpenTelemetry as well.
